middleware/rate_limiter_test.go

```diff
@@ -15,80 +19,106 @@ func TestRateLimiter(t *testing.T) {
 		return c.String(http.StatusOK, "test")
 	}
 
-	{
-		var inMemoryStore = new(RateLimiterMemoryStore)
-		inMemoryStore.rate = 1
-		inMemoryStore.burst = 3
-
-		assert.Panics(t, func() {
-			RateLimiter(nil)
-		})
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3)
 
-		assert.NotPanics(t, func() {
-			RateLimiter(inMemoryStore)
-		})
+	testCases := []struct {
+		id   string
+		code int
+	}{
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusTooManyRequests},
+		{"127.0.0.1", http.StatusTooManyRequests},
+		{"127.0.0.1", http.StatusTooManyRequests},
+		{"127.0.0.1", http.StatusTooManyRequests},
 	}
 
-	{
-		var skipped bool
-		var inMemoryStore = new(RateLimiterMemoryStore)
-		inMemoryStore.rate = 1
-		inMemoryStore.burst = 3
-
+	for _, tc := range testCases {
 		req := httptest.NewRequest(http.MethodGet, "/", nil)
-		req.Header.Add(echo.HeaderXRealIP, "127.0.0.1")
+		req.Header.Add(echo.HeaderXRealIP, tc.id)
 
 		rec := httptest.NewRecorder()
-
 		c := e.NewContext(req, rec)
-
-		mw := RateLimiterWithConfig(RateLimiterConfig{
-			Skipper: func(c echo.Context) bool {
-				skipped = true
-				return true
-			},
-			Store: inMemoryStore,
-			SourceFunc: func(ctx echo.Context) string {
-				return "127.0.0.1"
-			},
-		})
+		mw := RateLimiter(&inMemoryStore)
 
 		_ = mw(handler)(c)
+		assert.Equal(t, tc.code, rec.Code)
+	}
+}
+
+func TestRateLimiter_panicBehaviour(t *testing.T) {
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3)
+
+	assert.Panics(t, func() {
+		RateLimiter(nil)
+	})
+
+	assert.NotPanics(t, func() {
+		RateLimiter(&inMemoryStore)
+	})
+}
+
+func TestRateLimiterWithConfig(t *testing.T) {
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3)
+
+	e := echo.New()
 
-		assert.Equal(t, true, skipped)
+	handler := func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
 	}
 
-	{
-		var beforeRan bool
-		var inMemoryStore = new(RateLimiterMemoryStore)
-		inMemoryStore.rate = 1
-		inMemoryStore.burst = 3
+	testCases := []struct {
+		id   string
+		code int
+	}{
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusForbidden},
+		{"", http.StatusBadRequest},
+		{"127.0.0.1", http.StatusForbidden},
+		{"127.0.0.1", http.StatusForbidden},
+	}
 
+	for _, tc := range testCases {
 		req := httptest.NewRequest(http.MethodGet, "/", nil)
-		req.Header.Add(echo.HeaderXRealIP, "127.0.0.1")
+		req.Header.Add(echo.HeaderXRealIP, tc.id)
 
 		rec := httptest.NewRecorder()
 
 		c := e.NewContext(req, rec)
-
 		mw := RateLimiterWithConfig(RateLimiterConfig{
```

Comment:
Sorry for noticing this late, but the RateLimiterWithConfig should be the same instance for all testCases. That will be more representative of real usage and also the test will run faster because the instance is not created on every loop (I know I'm a little bit paranoid, there are only 7 tests 😞)
Please see that this is in other tests